packages/transaction-controller/src/TransactionController.ts

```diff
@@ -645,19 +704,21 @@ export class TransactionController extends BaseController<
       const chainId = parseInt(currentChainId, undefined);
       const { approved: status } = TransactionStatus;
 
-      const txNonce =
-        nonce ||
-        (await query(this.ethQuery, 'getTransactionCount', [from, 'pending']));
+      nonceLock = await this.nonceTracker.getNonceLock(from);
+
+      // if a nonce already exists on the transactionMeta it means this is a speedup or cancel transaction
+      // so we want to reuse that nonce and hope that it beats the previous attempt to chain. Otherwise use a new locked nonce
+      const nonceToUse = nonce ?? nonceLock.nextNonce;
```

Comment:
Maybe we avoid getting the nonce lock if there is already a nonce set? That might speed things up in some situations.